libraries/ms-drivers/src/stm32f0xx/spv1020_mppt.c

```diff
@@ -0,0 +1,57 @@
+#include "spv1020_mppt.h"
+#include "spv1020_mppt_defs.h"
+
+// This file implements stm32-specific functionality. See ../spv1020_mppt.c for the rest.
+
+// Send a command to SPV1020 and get a response.
+// It's safe to pass NULL as |rx_data| if |rx_len| is 0.
+static StatusCode prv_send_command(SpiPort port, uint8_t command, uint8_t *rx_data, size_t rx_len) {
+  if (port >= NUM_SPI_PORTS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Ensure rx_data can be NULL for rx_len==0 (spi_exchange should be safe, but this is more robust)
+  uint8_t placeholder;
+  if (rx_len == 0) {
+    rx_data = &placeholder;
+  }
+
+  // Note: it's possible that the SPV1020 uses 0x01 as the NOP instead of 0x00 as used by
+  // spi_exchange, in which case we need a weird custom SPI implementation to use 0x01.
+  // The command table gives the NOP command as 0x01, but the SPI implementaiton section says 0x00.
+
+  return spi_exchange(port, &command, 1, rx_data, rx_len);
+}
+
+StatusCode spv1020_shut(SpiPort port) {
+  return prv_send_command(port, COMMAND_SHUT, NULL, 0);
+}
+
+StatusCode spv1020_turn_on(SpiPort port) {
+  return prv_send_command(port, COMMAND_TURN_ON, NULL, 0);
+}
+
+StatusCode spv1020_read_current(SpiPort port, uint16_t *current) {
+  return prv_send_command(port, COMMAND_READ_CURRENT, (uint8_t *)current, 2);
+}
+
+StatusCode spv1020_read_voltage_in(SpiPort port, uint16_t *vin) {
+  return prv_send_command(port, COMMAND_READ_VIN, (uint8_t *)vin, 2);
+}
+
+StatusCode spv1020_read_pwm(SpiPort port, uint16_t *pwm) {
+  uint16_t raw_pwm;
+  StatusCode code = prv_send_command(port, COMMAND_READ_PWM, (uint8_t *)raw_pwm, 2);
```

Comment:
I believe this should be (uint8_t *)&raw_pwm, might need more brackets

Reply:
Whoops, fixed